mlflow/models/evaluation/utils/trace.py

```diff
@@ -0,0 +1,117 @@
+    is done, the original autologging configurations are restored.
+
+    Args:
+        enable_tracing (bool): Whether to enable tracing for the supported flavors during eval.
+    """
+    flavor_to_original_config = {}
+    trace_enabled_flavors = []
+
+    # AUTOLOGGING_INTEGRATIONS can change during we iterate over flavors and enable/disable
+    # autologging, therefore, we snapshot the current configuration to restore it later.
+    global_config_snapshot = AUTOLOGGING_INTEGRATIONS.copy()
+
+    for flavor in FLAVOR_TO_MODULE_NAME:
+        # TODO: Remove this once Gluon autologging is actually removed
+        if flavor == "gluon":
+            continue
+
+        try:
+            if autolog := _get_autolog_function(flavor):
+                original_config = global_config_snapshot.get(flavor, {}).copy()
+
+                # If autologging is explicitly disabled, do nothing.
+                if original_config.get("disable", False):
+                    continue
+
+                elif enable_tracing and _is_trace_autologging_supported(flavor):
+                    # set all log_xyz params to False except log_traces
+                    new_config = {
+                        k: False if k.startswith("log_") else v for k, v in original_config.items()
+                    }
+                    new_config |= {"log_traces": True, "silent": True}
+                    _kwargs_safe_invoke(autolog, new_config)
+                    trace_enabled_flavors.append(flavor)
+                elif flavor in AUTOLOGGING_INTEGRATIONS:
+                    # For flavors that does not support tracing, disable autologging
+                    autolog(disable=True)
+
+                flavor_to_original_config[flavor] = original_config
+
+        except Exception as e:
+            if isinstance(e, ImportError):
+                _logger.debug(f"Failed to import {flavor}. Skip updating autologging.")
+            else:
+                _logger.info(f"Failed to update autologging configuration for flavor {flavor}. {e}")
+
+            # Autologging configuration might be updated before the exception is raised,
+            # which needs to be reverted.
+            AUTOLOGGING_INTEGRATIONS.pop(flavor, None)
+
+    if trace_enabled_flavors:
+        _logger.info(
+            "Tracing is temporarily enabled during the model evaluation for computing some "
+            "metrics and debugging. To disable tracing, call `mlflow.autolog(disable=True)`."
+        )
+
+    try:
+        yield
+    finally:
+        # Restore original autologging configurations.
+        for flavor, original_config in flavor_to_original_config.items():
+            autolog = _get_autolog_function(flavor)
+            try:
+                if original_config:
+                    _kwargs_safe_invoke(autolog, original_config)
+                    AUTOLOGGING_INTEGRATIONS[flavor] = original_config
```

Comment:
Isn't this configuration for `AUTOLOGGING_INTEGRATIONS` done by autolog method itself?

Reply:
Great question. Ideally yes, but there is one exception; `globally_configured` key. Basically, when users call the general autolog `mlflow.autolog()` rather than each flavor's one and autologging is enabled by import hook, the config will have additional `"globally_configured": "true"` entry.

```
import mlflow
from mlflow.utils.autologging_utils import AUTOLOGGING_INTEGRATIONS

mlflow.autolog()

import sklearn
print(AUTOLOGGING_INTEGRATION.get("sklearn"))
```
```
{'log_input_examples': False, 'log_model_signatures': True, 'log_models': True, 'log_datasets': True, 'disable': False, 'exclusive': False, 'disable_for_unsupported_versions': False, 'silent': False, 'max_tuning_runs': 5, 'log_post_training_metrics': True, 'serialization_format': 'cloudpickle', 'registered_model_name': None, 'pos_label': None, 'extra_tags': None, 'globally_configured': True}
```

This flag is removed when we then call the flavor's autolog function.
```
mlflow.sklearn.autolog()
print(AUTOLOGGING_INTEGRATION.get("sklearn"))
```
```
{'log_input_examples': False, 'log_model_signatures': True, 'log_models': True, 'log_datasets': True, 'disable': False, 'exclusive': False, 'disable_for_unsupported_versions': False, 'silent': False, 'max_tuning_runs': 5, 'log_post_training_metrics': True, 'serialization_format': 'cloudpickle', 'registered_model_name': None, 'pos_label': None, 'extra_tags': None}
```
(the new config doesn't include `globally_configured`)

This flag seems to play an important role. For example, if this flag is set to `True`, `mlflow.autolog(disable=True)` does not override the flavor's configuration. Therefore, we need to explicitly set the configuration to restore this flag.

Reply:
I see, let's refactor this type of ad-hoc logic of autologging. This seems hard to maintain, as we dicscussed today.

Reply:
Yea, created a JIRA. I think this gonna be one of the most cumbersome refactoring; number of edge cases, environment differences, and lots of inter-test side effect due to global state....